girdermedviewer/app/girder/utils.py

```diff
@@ -14,55 +14,80 @@ class CacheMode(Enum):
+                raise Exception("A directory must be provided if cache mode is Permanent")
+            self.temporary_directory = TemporaryDirectory()
+            self.temp_dir_path = self.temporary_directory.name
+
+        if self.assetstore_dir_path is not None:
+            if (
+                os.path.normcase(os.path.realpath(os.path.abspath(self.assetstore_dir_path))) ==
+                os.path.normcase(os.path.realpath(os.path.abspath(self.temp_dir_path)))
+            ):
+                raise Exception("The temporary directory cannot match the assetstore directory.")
+
     def __del__(self):
         if self.cache == CacheMode.Session:
             self.clear_cache()
 
+    def _download_file(self, file, file_path):
+        logger.info(f"Download {file['name']} to {file_path}")
+        self.girder_client.downloadFile(
+            file["_id"],
+            file_path
+        )
+
     def get_item_files(self, item):
         return self.girder_client.listFile(item["_id"])
 
     @contextmanager
-    def download_file(self, file):
+    def fetch_file(self, file):
         """
-        :param forced_cache can be used to override the cache mode for a specific file
+        First check if `file` does not already exist in assetstore.
+        Then check if it does not already exist in cache.
+        Finally download it if needed
         """
-        file_path = os.path.join(self.temp_dir_path, file["_id"], file["name"])
-        if not os.path.exists(file_path):
-            logger.info(f"Download {file['name']} to {file_path}")
-            self.girder_client.downloadFile(
-                file["_id"],
-                file_path
-            )
+        file_path = None
+        if self.assetstore_dir_path is not None:
+            if "path" not in file:
+                raise Exception("The Girder file is missing 'path' information. Make sure to use the girdermedviewer-plugin")
+            file_path = os.path.join(self.assetstore_dir_path, file['path'])
+            if not os.path.exists(file_path):
+                logger.warning(f"The file {file_path} cannot be read from the assetstore, it will be downloaded instead")
+                file_path = None
+
+        if file_path is None:
+            file_path = os.path.join(self.temp_dir_path, file['_id'], file["name"])
+            if not os.path.exists(file_path):
+                self._download_file(file, file_path)
+
         try:
             yield file_path
         finally:
             if self.cache == CacheMode.No:
-                self.clear_cache(file)
+                self.clear_cache(file_path)
 
-    def clear_cache(self, file=None):
-        if file is not None:
-            file_path = os.path.join(self.temp_dir_path, file['_id'], file["name"])
-            os.remove(file_path)
+    def clear_cache(self, file_path=None):
+        if file_path is not None:
+            if os.path.exists(file_path):
+                os.remove(file_path)

```

Comment:
When I said [that](https://github.com/KitwareMedical/GirderMedViewer/pull/18#discussion_r1949865713), it was when `clear_cache` was only deleting files in `temp_dir_path`.
The way you wrote it here, it may delete the file from the assetstore.